### DISCUSSION TITLE
I’m trying to create a DeepVAREstimator model and I keep running into a error

### DISCUSSION THREAD

**Discussion by humzaharoon** [awslabs/gluonts#1664] [Q&A] [answered]
The following is the code I wrote so far for my gluonts deepvar model. I'm trying to use multiple time series to make a prediction and gluonts seemed to be a good option for that. The following below is the code of what I've done so far and the error:

`import  numpy as np   
import  pandas as pd
import  matplotlib.pyplot  as plt
from gluonts.model.deepvar import DeepVAREstimator
from gluonts.mx.distribution import GaussianOutput
from gluonts.mx.trainer import Trainer
from sklearn.preprocessing import MinMaxScaler
from statsmodels.tsa.stattools import grangercausalitytests

df = pd.read_csv('data_share3.csv', index_col=0, header=0, parse_dates=True)

df.iloc[:200, 0].plot(figsize=(12, 5), linewidth=2)
plt.grid()
plt.show()

from gluonts.dataset.common import ListDataset

training_data = ListDataset(
    [{'start':df.index[i], 
      'target': tuple(t),
     } for t in df.values[:200, :]
       for i in range(200)],
    freq="3M"  
)
testing_data = ListDataset(
    [{'start': df.index[i], 'target': t}  
     for t in df.values[200:227, :]
     for i in range(200, 227)],
    freq='3M'
)

prediction_length = 12

estimator = DeepVAREstimator(freq='3M',
                            target_dim = 4000, 
                           prediction_length = prediction_length,
                           context_length = 145,
                           num_layers=2,
                           num_cells=50,
                           cell_type='lstm',
                           distr_output = GaussianOutput(),
                           dropout_rate = 0.00005, 
                           trainer=Trainer(epochs=10, 
                            learning_rate=1E-3, 
                            hybridize = True, 
                            batch_size=245,  
                            num_batches_per_epoch = 12))

predictor=estimator.train(training_data) `

The error I keep getting is below:

`---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-145-99705a50d715> in <module>
----> 1 predictor=estimator.train(training_data)

/opt/anaconda3/lib/python3.8/site-packages/gluonts/mx/model/estimator.py in train(self, training_data, validation_data, num_workers, num_prefetch, shuffle_buffer_length, cache_data, **kwargs)
    193         **kwargs,
    194     ) -> Predictor:
--> 195         return self.train_model(
    196             training_data=training_data,
    197             validation_data=validation_data,

/opt/anaconda3/lib/python3.8/site-packages/gluonts/mx/model/estimator.py in train_model(self, training_data, validation_data, num_workers, num_prefetch, shuffle_buffer_length, cache_data)
    142         cache_data: bool = False,
    143     ) -> TrainOutput:
--> 144         transformation = self.create_transformation()
    145 
    146         transformed_training_data = transformation.apply(training_data)

/opt/anaconda3/lib/python3.8/site-packages/gluonts/model/deepvar/_estimator.py in create_transformation(self)
    340                 ExpandDimArray(
    341                     field=FieldName.TARGET,
--> 342                     axis=0 if self.distr_output.event_shape[0] == 1 else None,
    343                 ),
    344                 AddObservedValuesIndicator(

IndexError: tuple index out of range `

The above is the error I keep receiving. I would greatly appreciate someone's input.

**Comment by mbohlkeschneider** [comment]
As this an error in the transformation chain, it is hard to debug without the data. Can you add a synthetic data snippet that produces the error?

**Reply by humzaharoon** [reply]
I receive the error when I try to run the model. The following is the data snippet along with the data frame its in: 
<img width="934" alt="Screen Shot 2021-08-26 at 9 30 25 PM" src="https://user-images.githubusercontent.com/73046458/131057253-144eceaf-7ef9-49d9-9227-db40f4355edc.png">
<img width="737" alt="Screen Shot 2021-08-26 at 9 31 25 PM" src="https://user-images.githubusercontent.com/73046458/131057328-94e9cc92-43ad-43b8-94d3-1e68e7356638.png">
<img width="755" alt="Screen Shot 2021-08-26 at 9 31 44 PM" src="https://user-images.githubusercontent.com/73046458/131057348-569a337c-6cbe-488c-a36c-ea6dded2a427.png">

**Comment by mbohlkeschneider** [comment, marked as answer]
Thank you for posting the data. 

A few clarifications: 

1. As far as I can tell from your data, the target is a 1D array. DeepVAR is a multivariate model and will expect a 2D target input of shape `num_time_steps * target_dim`. 
2. When loading a 2D target, you need to set the flag `one_dim_target` in the dataset constructor to `False`. 
3. Lastly, because DeepVAR is multivariate, you would want to take an actual multivariate distribution like `MultivariateGaussianOutput` or `LowrankMultivariateGaussianOutput`. 

Here is a snippet that should run and summarizes the steps above: 

```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from gluonts.model.deepvar import DeepVAREstimator
from gluonts.mx.distribution import MultivariateGaussianOutput
from gluonts.mx.trainer import Trainer

from gluonts.dataset.common import ListDataset

training_data = ListDataset(
[
    {'start':"2020-01-01",
    'target': [[1]*100, [2]*100]}],freq="3M", one_dim_target=False)

prediction_length = 12

estimator = DeepVAREstimator(freq='3M',
target_dim = 2,
prediction_length = prediction_length,
context_length = 12,
num_layers=2,
num_cells=50,
cell_type='lstm',
distr_output = MultivariateGaussianOutput(dim=2),
dropout_rate = 0.00005,
trainer=Trainer(epochs=10,
learning_rate=1E-3,
hybridize = True,
batch_size=245,
num_batches_per_epoch = 12))

predictor=estimator.train(training_data)
```

**Reply by humzaharoon** [reply]
@mbohlkeschneider  Thank you so much for your help, I really appreciate it. I just had a question regarding the target, you are passing in 'target': [[1]*100, [2]*100]}. I understand that this is 2 Dimensional, but are you multiplying by 100 because those are the time steps and [1] and [2] are rows and columns? I have tried to create the following for loop: 

`[{'start':df.index[i], 
      'target': t,
     } for t in df.values[:200, :]
`
How do I implement a 2D data structure?
